README.md

# **README on Mcalibrator2**

<div>Created    : "2013-05-13 12:30:52 ban"</div>
<div>Last Update: "2018-02-03 11:29:29 ban"

**********
# <a name = "Menu"> **Menu** </a>

- [Introduction – what is Mcalibrator2 ?](#Introduction)
- [System requirements](#System)
- [How to launch Mcalibrator2](#Launch)
- [How to generate/update html-based Mcalibrator2 help files](#Help)
- [Detailed manual of Mcalibrator2](#Manual)
  - [Aims and purposes](#Purposes)
  - [Installation](#Installation)
  - [Mcalibrator2 configuration - setting measurement/calibration parameters](#Configuration)
  - [Measure tab (you can measure CIE1931 xyY here)](#Measure)
  - [LUT tab - Running Gamma-correction and generating Color-LookUp tables (CLUTs)](#LUT)
  - [Chromaticity (CIE1931 xyY etc) estimation procedure](#Chromaticity)
  - [About tab](#About)
  - [How to make Mcalibrator2 communicate with your own colorimeter(s) or photometer(s)](#CustomColorimeter)
  - [How to add your own chromaticity estimation methods to Mcalibrator2](#CustomEstimation)
  - [Automations and scripting of Mcalibrator2](#Scripting)
- [Acknowledgments](#Acknowledgments)
- [License](#License)
- [Citations](#Citations)
- [TODO](#TODO)

**********

# <a name = "Introduction"> **Introduction – what is Mcalibrator2 ?** </a>

![Mcalibrator2](doc/imgs/00_mc_about.png)    ![Mcalibrator2](doc/imgs/00_mc_select.png)  

Mcalibrator2 is a ***MATLAB***-based GUI display luminance and chromaticity characterization software package for visual neuroscience and psychology studies. The package is especially focusing on  
1. providing accurate gamma-correction, and  
2. finding the best RGB video inputs to produce the required CIE1931 xyY values using several goal-seeking optimization algorithms.  

(Matlab is a registered trademark of [***The Mathworks Inc.*** ](https://www.mathworks.com/) )

For details, please read documents in *~/Mcalibrator2/doc* directory.
Also please see the link below.

- [**Mcalibrator2 original website**](http://www.cv.jinkan.kyoto-u.ac.jp/site/mcalibrator/)
- [**Mcalibrator2 GitHub repository**](https://github.com/hiroshiban/Mcalibrator2)

The details of the algorithms we developed are described below.
- [**Original paper published in Journal of Vision, 2013**](http://www.journalofvision.org/content/13/6/20.long)

Thank you for using our software.  
We are happy if Mcalibrator2 can help your research projects.

[return to menu](#Menu)

# <a name = "System"> **System requirements** </a>

- **OS: Windows XP/VISTA/7/8/10 or Mac OSX**
  - note 1: Windows OS is required to use some photometers with Mcalibrator2 as only Windows drivers are distributed for them.
  - note 2: On a Mac OSX box, an interface to a virtual serial port is required (generally implemented by default).
  - note 3: We have not tested Mcalibrator2 on any Linux boxes. But it would work once you can get a proper serial or usb interface with a valid driver to control your photometer.

- **MATLAB R2009a** or later (We have tested the latest version with **R2014b** and **R2017a**), and "Optimization" and "Statistics" toolboxes
  - note 1: Mcalibrator2 works even without optimization/statistics toolboxes, but to use full functionality of the software (e.g. automatic non-linear RGB video input estimations), these toolboxes are required.
  - note 2: Mcalibrator2 can not be launched properly on MATLAB R2007 or older since the GUI handling system that we are using to control Mcalibrator2 windows is quite different from the conventions of the older MATLABs.

[return to menu](#Menu)

# <a name = "Launch"> **How to launch Mcalibrator2** </a>

To launch Mcalibrator on MATLAB, please run

```Matlab
>> Mcalibrator2
```

Please note that when you use MATLAB version 2012 or later, a function to control GUI tabs, "tabselectionfcn", will be disabled due to incompatibility of the function with the recent MATLAB.

[return to menu](#Menu)

# <a name = "Help"> **How to generate/update html-based Mcalibrator2 help files** </a>

Please run

```Matlab
>> update_Mcalibrator2_html_docs
```

Then, all the html-based help documents will be automatically generated in **~/Mcalibrator2/doc/html**.  
To read the help documents, please open **~/Mcalibrator2/doc/html/index.html** on your browser.

[return to menu](#Menu)

# <a name = "Manual"> **Detailed manual of Mcalibrator2** </a>

# <a name = "Purposes"> Aims and purposes </a>

Display characterization is an essential part of the experimental procedures in vision science because almost all current experiments are conducted in computer-aided environments: visual stimuli are manipulated via a programming language and displayed on a computer display, and observer's responses are acquired by pressing keys connected to a computer. To ensure that visual stimuli - their luminance, color, timing, etc. are presented precisely in such computer-based experiments, researchers need to characterize display devices accurately in advance of actual experiments.  

CRT (Cathode Ray-Tube) displays are the most widely-used devices for current vision experiments, and the calibration procedures to characterize their luminance and chromaticities are well-established in the two-stage procedures: gamma-correction, followed by a linear color transformation (Berns et al., 1996; Brainard et al., 2002; Ban et al., 2006). The calibration results obtained through this standard two-step procedure has been tested (Brainard et al., 2002; Ban et al., 2006) and the quality of luminance and chromatic stimuli on CRT displays satisfy the researchers' criterion.  

However, non-CRT devices such as a LCD (Liquid Crystal Display) and a DLP (Digital Light Processing) have come into the main stream recently and researchers are required to use non-CRT over CRT devices. There is no evidence that the standard display characterization method can be applied to these new types of devices since the current widely-used standard display calibration method is established based on the internal model of CRT devices.  

We have therefore developed a non-device-specific approach to display characterization. Specifically, our new methods use model-free gamma-correction procedure combined with a linear/non-linear (Nelder-Mead Simplex, Nelder & Mead, 1965; Dennis & Woods, 1987) hybrid or line search (Powell's method with Coggins constrain, Powell, 1964; Brent, 1973; Press et al., 2007; Farhi, 2011; Farhi et al, 2012) algorithm to get the optimal RGB video input values to produce the required luminance and chromaticities. The methods only assume 1) a monotonic increment of luminance against the increment of video input values, and 2) a piece wise linearity of the system in the initial estimation step. These new methods have a much broader range of applicability because they do not presume the internal model of the display device and can handle non-linearity of the device.  

The whole procedures are integrated into a GUI-based display characterization software written in MATLAB (The Mathworks, Inc, USA) language and termed "Mcalibrator2" (Ban & Yamamoto, 2013). The applicability and efficiency of our software to a wide range of display devices, including LCD and DLP type of devices, were confirmed by comparing the calibration accuracies of our procedures with that of the standard two-stage method. We found that all of these new approaches improved calibration accuracies for non-CRT devices compared with the standard display characterization procedure.
Mcalibrator2 has provided some additional components to communicate with any photometers that researchers are using in their daily measurements. The software has also provided simple frameworks to add alternative chromaticity optimization procedures researchers may want to use. Further, the software can automatically generate gamma-correction tables compatible with Psychtoolbox, one of the most widely used vision science tools (Brainard, 1997; Pelli, 1997). These functions will assist researchers in characterizing their display devices.  

The authors provide this software suite in the hope that it will benefit researchers performing calibration of their display devices efficiently and improve accuracies of stimulus displays regardless of the display types.

![Mcalibrator2 schematic illustrations of chromaticity estimation procedures](doc/imgs/00_Schematic_illustration_of_chromaticity_estimations.png)  
**Schematic illustrations of the Gamma-correction and chromaticity estimation procedures implemented in Mcalibrator2.**  
a) Standard Gamma-correction procedures. b) Standard linear chromaticity estimation procedure. c) Custom Gamma-correction procedures. d) Custom chromaticity estimation procedures based on goal-seeking optimization algorithms  

***References***

1.	Ban, H., Yamamoto, H., Ejima, Y. (2006). Mcalibrator: MATLAB integrated GUI software for display calibration ― a proposal of a new color calibration procedure applicable to a wide range of display devices and evaluation of its efficiency ―. The Japanese Journal of Psychonomic Science. 24(2). 149-161.
2. Ban, H., & Yamamoto, H. (2013). A non-device-specific approach to display characterization based on linear, nonlinear, and hybrid search algorithms. Journal of Vision, 13(6):20, 1-26.
3.	Berns, R.S. (1996). Methods for Characterizing CRT Displays. Displays. 16, 173-182.
4.	Brainard, D.H. (1997). The psychophysics toolbox. Spatial Vision. 10, 433-436.
5.	Brainard, D.H., Pelli, D.G., Robson, T. (2002). Display Characterization. Encyclopedia of Imaging Science and Technology. Wiley, 172-188
6.	Brent, R.P. (1973). Section 7.3: Powell's algorithm. In Algorithms for minimization without derivatives. Englewood Cliffs, N.J.: Prentice-Hall.
7.	Dennis, J.E.Jr., Woods, D.J. (1987). New Computing Environments: Microcomputers in Large-Scale Computing. Edited by A. Wouk, SIAM, 116-122.
8.	Farhi, E. (2011). The iFit data analysis library, ILL, Computing for Science. Online reference: *http://ifit.mccode.org*
9.	Farhi, E., Debab, Y., Willendrup, P. (in press). Journal of Neutron Research.
10.	Nelder, J.A., Mead, R. (1965). A Simplex Method for Function Minimization. Computer Journal, 7, 308-313.
11.	Pelli, D.G. (1997). The video toolbox software for visual psychophysics: transforming numbers into movies. Spatial Vision. 10, 437-442.
12.	Powell, M. J. D. (1964). An efficient method for finding the minimum of a function of several variables without calculating derivatives. Computer Journal 7 (2): 155-162.
13.	Press, W.H., Teukolsky, S.A., Vetterling, W.T., Flannery, B.P. (2007). Section 9.3. Van Wijngaarden–Dekker–Brent Method. Numerical Recipes: The Art of Scientific Computing (3rd ed.). New York: Cambridge University Press.

[return to menu](#Menu)

# <a name = "Installation"> Installation </a>

1.	You need to install **MATLAB** ([***The Mathworks Inc.*** ](https://www.mathworks.com/)) to use Mcalibrator2.  
    You may additionally need to install **"Optimization"** and **"Statistics"** toolboxes to use all the functions of Mcalibrator2.
2.	Please get the latest version of Mcalibrator2 from the link below,  
    [**Mcalibrator2 GitHub repository**](https://github.com/hiroshiban/Mcalibrator2)  
    and add the *~/Mcalibrator2* directory to your MATLAB-path. If you use a MATLAB after 2011a, you do not need to add subdirectories.  
    Mcalibrator2 can handle all the subrouines internally.  
    If you use MATLAB before 2011a, please add the Mcalibrator2 directory including all the subdirectories into your MATLAB-path.  

3.  Please launch Mcalibrator2 from the MATLAB console,  
    
    ```Matlab
    >> Mcalibrator2
    ```

    Then, you will see the main window of Mcalibrator2 as below.

[return to menu](#Menu)

# <a name = "Configuration"> Mcalibrator2 configuration - setting measurement/calibration parameters </a>

![Mcalibrator2 config tab](doc/imgs/01_config_tab.png)  
**Mcalibrator2: config tab**

The "config" tab above is for setting measurement/calibration parameters.  

1.  You can set all the measurement/calibration parameters in this tab.  
    The details of the configuration parameters are described on the "Info" window.  
    After setting all the parameters, please press "OK" button.
    Then, all the parameters are saved and "measure", "LUT", "Color Calculator" tabs will be ready to be used.
2.	If you press "Save config" button, the current parameters will be saved under ***~/Mcalibrator2/config*** directory.  
    Yuo can load those parameters anytime with pressing "Load config" button.
3.	On the **[date]** editing window, you can set the data name as you like such as *yymmdd_1*, *yymmdd_2*, or *yymmdd_display1*, as well as *yymmdd* format.  
    When you need to calibrate a display device twice or more in a day, please change [date] names so that they are not overlapped.
4.	When you set a previous date on the **[date]** window, and if the corresponding data directory exists in ***~/Mcalibrator2/data***, you can load those previous data on memory.  

Next, you can proceed to measure RGB phosphor luminance values. Please go to "measure" tab.

[return to menu](#Menu)

# <a name = "Measure"> Measure tab (you can measure CIE1931 xyY here) </a>

![Mcalibrator2 measure tab](doc/imgs/02_measure_tab.png)  
**Mcalibrator2: measure tab**

You can set up a photometer and measure luminance values on this ***"measure"*** tab.  

First, please press "Adjust Colorimeter Position" button. Then, the grid window will be displayed on the screen. You can adjust and set a colorimeter focus using this window. The colorimeter position can be adjusted and validated anytime when you press the ***"Adjust Colorimeter Position"*** button. 

![Mcalibrator2 measurement position adjustment](doc/imgs/03_adjust_window.png)  
**Mcalibrator2: adjustment of your colorimeter position**

Next, set an ID of the serial (or USB) port to which your colorimeter is connected. On Windows, the ID will be like ***COM1***, ***COM2***, ..., on Linux, it will be like ***/dev/ttyS0***, and on Mac, it will be like ***/dev/tty.KeySerial1***.  
(***Please note that the authors have mainly tested Mcalibrator2 on Windows PC. We have not tested the software on Mac or Linux machines. Sorry***.)  
After setting the ID, please press ***"Create"*** button, and a serial (or USB) object to communicate with the colorimeter will be generated. Then, by pressing ***"Initialize Apparatus"*** button, the colorimeter will be initialized through the port.

The initial setups finished now. If you press ***"Measure CIE1931 xyY"*** button, the luminance values are measured against 32 (by default) video inputs for each of the RGB phosphors.  

After completing the measurements, the relationship between video input and luminance output values will be plotted as below.  

![Mcalibrator2 measure tab after measurements](doc/imgs/04_measure_tab_after_measurements.png)  
**Mcalibrator2: measure tab after measurements**

The whole procedure on "measure" tab is completed now.  
Next, Mcalibrator2 is going to generate Color Lookup Tables (CLUTs) by linearizing the input/output relationship. Please go to ***"LUT"*** tab.

[return to menu](#Menu)

# <a name = "LUT"> LUT tab - Running Gamma-correction and generating Color-LookUp tables (CLUTs) </a>

![Mcalibrator2 LUT tab](doc/imgs/05_LUT_tab.png)  
**Mcalibrator2: LUT tab (performing gamma-correction and generating CLUTs)**

1.	Please select a fitting model to describe the input/output relationship.  
    - When you are using CRTs, ***"Gain-Offset-Gamma"*** (a well-known standard gamma-correction model) is recommended.
    - When you are using a LCD or DLP, ***"cubic spline"*** or ***"robust cubic spline"*** is recommended.  
    - If you select ***"grid search"*** , LUTs are generated by directly searching the best values,  
      but it will only work when you measure all 256 points (when you set #LUT step to 256 on an 8-bit phosphor).

![Mcalibrator2 LUT tab model selection](doc/imgs/05_LUT_tab_model_selection.png)  
**Mcalibrator2: Selecting a model for a gamma-correction**

2.	Please press ***"Fit a model"*** button, the selected model is fitted to the data and the results will be displayed on the ***"measure"*** tab.

![Mcalibrator2 measure tab after the gamma correction](doc/imgs/06_measure_tab_after_gamma_correction.png)  
**Mcalibrator2: Gamma-correction results**

3.	If the fitting results are fine, please press ***"Create Color Lookup Table"*** button.  
    CLUTs will be generated and saved to ***~/Mcalibrator2/data/(date etc)*** .
4.	The details of the measurements and the model fits (input/output relationship before/after gamma-corrections, LUTs, residuals of the fits etc.) will be plotted on the external figure windows. If you are running Mcalibrator2 on Windows in which ***Microsoft(C) PowerPoint*** is installed, these plots are saved into a PowerPoint slide file and stored to ***~/Mcalibrator2/data/(date etc.)*** automatically.
5.	You can repeat LUT-generation procedure by changing a fitting model until you can get the best results.

![Mcalibrator2 LUT tab after generating CLUTs](doc/imgs/07_LUT_tab_after_generating_CLUTs.png)  
**Mcalibrator2: LUT tab after generating CLUTs**

![Mcalibrator2 model fit result](doc/imgs/08_model_fit_result.png)  
**Mcalibrator2: Model fit result**

6.	To check how properly the Gamma-correction is performed, please press the ***"Check the linearity"*** button. When the ***"easy check"*** toggle button is ON (default), the generated CLUTs will be simply displayed on LUT tab. If it is set to OFF, luminance values for 20 video inputs (equally sampled from 0.0 to 1.0) are re-measured, and the linearity between the input video values and the output luminance are evaluated. If the results are not good, please re-generate CLUTs by changing a fitting model.  

![Mcalibrator2 linearity test](doc/imgs/09_LUT_tab_linearity_test.png)  
**Mcalibrator2: Linearity test (Easy-check mode)**

![Mcalibrator2 linearity test](doc/imgs/10_LUT_tab_linearity_test_remeasured.png)  
**Mcalibrator2: Linearity test (Remeasuring mode)**

The whole procedures of gamma-correction and CLUT generations are completed. If you do not need to get optimal RBG video inputs to produce the specific chromaticities (CIE1931 xyY, not only Y (luminance)), the display characterization is done now.

If you need to perform chromaticity calibration, please go to ***"Color Calculator"*** tab below. 

[return to menu](#Menu)

# <a name = "Chromaticity"> Chromaticity (CIE1931 xyY etc) estimation procedure </a>

![Mcalibrator2 Color Calculator](doc/imgs/11_Color_Calculator_tab.png)  
**Mcalibrator2: Color Calculator tab (for performing chromaticity calibration)**

1. First, to estimate chromaticity by a simple linear transformation, please load CIE1931 xyY values for each of the RGB phosphors at their maximum levels. To do this, please press the ***"Load/measure RGB"*** button. The phosphor values measured/stored in the previous session are extracted and displayed on the CIE1931 chromaticity diagram as 3 circles and a triangle. The range of chromaticity values you can present with the current display is a region enclosed by this triangle.

![Mcalibrator2 Color Calculator after loading phosphors](doc/imgs/12_Color_Calculator_tab_after_loading_phosphors.png)  
**Mcalibrator2: Color Calculator tab after loading CIE1931 xy values of the RGB phosphors at their maximum level.**

2. Then, if you want to use the CLUTs generated in the previous step, please set the ***LUT*** radio-button to ON (default). If you want to use the raw RGB values without the reference CLUTs, please set the ***RGB*** radio-button to ON. If the RGB radio-button is ON, you can directly set RGB values on the ***"you want to measure"*** edit box. Please use this option when you want to simply measure RGB values of your stimuli.
3. Next, please input CIE1931 xyY values you want to use in your experiment on ***"x,y,Y"*** editing window just below the ***"you want (x1, y1, Y1; x2, y2, Y2; x3, y3,..)"*** message box. Please describe 3 (x, y, Y) values with a comma (,) as a delimiter. If you want to set multiple values, please use a semi-colon (;) to separate each of CIE 1931 xyY values. The format is thus as below.
**x1,y1,Y1; x2,y2,Y2; x3,y3,Y3;...**
4. You can also load these CIE1931 xyY values from a text file. To do this, please press the ***"load"*** button. The CIE1931 xyY values in the text file should be described with commas and semi-colons as delimiters as below.  
***x1, y1, Y1;***  
***x2, y2, Y2;***  
***x3, y3, Y3;***  
***...***  
5. After setting xyY values you want, please press ***"convert xyY"*** button. If the LUT radio-button is ON, the xyY values are converted to **CLUT IDs**. If the RGB radio-button is ON, the raw RGB values required are calculated by a simple linear transformation. The estimated values  are presented on the ***"LUT (or RGB)"*** edit box on the right side.
6. To measure actual CIE1931 xyY values for these initial estimations, please press the ***"measure"*** button. The chromaticities for the estimated RGB values are successively measured. The results are presented on the ***"x,y,Y"*** and ***"LUT(or RGB)"*** edit boxes just below the ***"results (x1, y1, Y1; x2, y2, Y2; x3, y3,..)"*** message box. In addition, the CIE1931 xy values you want and the actual measures for the estimations are plotted on the ***"CIE1931 chromaticity diagram"*** window by a circle (what you want) and a cross (actual measurement). If these two values are close enough, the chromaticity calibration is done successfully.

![Mcalibrator2 Color Calculator tab single estimation](doc/imgs/12_Color_Calculator_tab_single_estimation.png)  
**Mcalibrator2: Color Calculator tab, single color estimation result**

7. If the estimations are not good, you need to adjut ***LUT IDs*** or ***RGB*** values by an automatic chromaticity estimation/correction method.
- Mcalibrator2 currently has 7 automatic chromaticity estimation methods;
  1. ***recursive linear estimations based on a least-squares method with random local-value samplings***,  
  2. ***recursive linear estimations based on a least-squares method with grid local-value samplings***,
  3. ***adjusting residuals by linear coeeficient estimations***,  
  4. ***a non-linear direct search algorithm --- Nelder-Mead Simplex***,  
  5. ***a linear/nonlinear hybrid estimation --- recursive linear estimations + Nelder-Mead Simplex***,  
  6. ***a line search algorithm --- Brent-Powell with Coggins/Golden-section***,  
  7. ***a linear/nonlinear hybrid estimation --- recursive linear estimations + Brent-Powell with Coggins/Golden-section***.  
  For more details of the estimation algorithms, please see comments on the source codes of Mcalibrator2 and see the paper on this software ([Ban and Yamamoto, 2013, Journal of Vision](http://www.journalofvision.org/content/13/6/20.long)). You can select one of these methods by using the dropdown-box next to the ***"measure"*** button. The estimation for one chromaticity will take 1-5 min. After finishing the estimations, the results are presented on the ***"x,y,Y"*** and ***"LUT(or RGB)"*** edit boxes just below the ***"results (x1, y1, Y1; x2, y2, Y2; x3, y3,..)"*** message. In addition, the CIE1931 xy values you want and the actual measured values for the estimations are plotted on the ***"CIE1931 chromaticity diagram"*** window by a circle (what you want) and a cross (actual measurement).

![Mcalibrator2 Color Calculator tab selection of estimation methods](doc/imgs/13_Color_Calculator_tab_selection_of_estimation_methods.png)  
**Mcalibrator2: Color Calculator tab, the dropdown-list for selecting a chromaticity estimation method**

![Mcalibrator2 Color Calculator tab automatic color estimations](doc/imgs/14_Color_Calculator_tab_automatic_color_estimations.png)  
**Mcalibrator2: Color Calculator tab, results of automatic chromaticity calibration (50 chromaticities)**

8. To view the detailed chromaticity estimation values, please press the ***"view"*** button. The estimation results, residuals etc. are displayed on the MATLAB console window.
9. To save the estimation results, please press the ***"save"*** button. The results are saved as a *.mat file to ***~/Mcalibrator2/data/(date etc.)*** like ***"estimate_files_001.txt"*** , ***"estimate_files_002.txt"*** ,... (file prefixex _001, _002,... are added automatically). Please be careful that ***no result will be saved automatically*** without pressing the ***"save"*** button.
10. If you can reproduce chromaticity values you want accurately enough, all the calibration procedures are completed. If the results are not good, please retry the estimations using different methods.

[return to menu](#Menu)

# <a name = "About"> About tab </a>

![Mcalibrator2 About tab](doc/imgs/15_about_tab.png)  
**Mcalibrator2: about tab**

![Mcalibrator2 display test](doc/imgs/16_display_test.png)  
**Mcalibrator2: display test**

![Mcalibrator2 color flickering test](doc/imgs/17_color_flickering_test.png)  
**Mcalibrator2: color flickering test**

On the ***"About"*** tab, the details of Mcalibrator2 software packages are displayed. If you press the ***"display test"*** button on this tab, you can perform several simple display tests.  
Furthremore, by pressing the ***"parameters"*** button, you can change the detailed gamma-correction, LUT generation, and chromaticity estimation parameters as you like. ***However, changing those default values affect the performance and accuracies of the Mcalibrator2 functions. Please be careful.***

[return to menu](#Menu)

# <a name = "CustomColorimeter"> How to make Mcalibrator2 communicate with your own colorimeter(s) </a>

Currently, we have already prepared subrountines for Mcalibrator2 to communicate with  
- *Konica-Minolta CS-100A (tested),*  
- *Admesy Brontes-LL (tested),*  
- *Photo-research PR-650 (tested),*  
- *Cambridge Research Systems ColorCal2 (tested)*  
- *some more but NOT fully tested.*

You can also use your own colorimeter(s) with Mcalibrator2 if you write some simple codes in MATLAB. To do this, you need to  
1.Register your own colorimeter(s) to an apparatus-list file, and  
2.Create a class file which defines a MATLAB object for your own colorimeter to communicate with Mcalibrator2.

To register your colorimeter, please edit ***~/Mcalibrator2/subfunctions/colorimeter/colorimeter_list.m***
By default, several photometers are already registered as below in this file.

```Matlab
colorimeters{1}={'Photo Research PR-650','pr650',0};
colorimeters{2}={'KONICA-MINOLTA CS-100A','cs100a',0};
colorimeters{3}={'KONICA-MINOLTA CS-150','cs150',1};
colorimeters{4}={'Admesy Brontes-LL (64bit)','brontesLL',1};
colorimeters{5}={'Admesy Brontes-LL (32bit)','brontesLL32',1};
colorimeters{6}={'Cambridge Research Systems ColorCAL2 Win-USB','colorcal',1};
%colorimeters{7}={'Cambridge Research Systems ColorCAL2 (PTB, not tested)','colorcal2',1};
colorimeters{7}={'Cambridge Research Systems ColorCAL CDC (Virtual Port)','colorcalcdc',0};
colorimeters{8}={'Cambridge Research Systems OptiCal (not tested)','optical',0};
colorimeters{9}={'International Light IL1700','il1700',0};
colorimeters{10}={'Dummy Colorimeter (for Debug)','dummy_colorimeter',0};
```

Therefore, please add your colorimeter(s) as ***colorimeters{11}*** , ***colorimeters{12}*** ,...or you can fully re-write this list.  

doc/history.txt
Mcalibrator2 --- MATLAB software for display luminance/color characterization

Copyright (c) 2013, Hiroshi Ban, Kyoto University, and
National Institute of Information and Communications Technology, Japan.

Created    : "July 01 2012 2:26:30 ban"
Last Update: "2018-02-01 20:20:54 ban"

==================== History ====================
July 01 2012 Mcalibrator2 ver.1.0 released. The first stable version of Mcalibrator2.
May  31 2013 Mcalibrator2 paper by H.Ban & H.Yamamoto was published in Journal of Vision (2013) 13(6).20, 1-26. doi: 10.1167/13.6.20
June 09 2013 Mcalibrator2 ver.1.1 released. Compatible with Mac OSX10.8.3 and Windows 8 professional 64bit with MATLAB2013a
Dec  11 2013 Mcalibrator2 ver.1.2 released. Several modifications. For details, please see .git revision log
Dec  12 2013 Mcalibrator2 was registered to GitHub
Feb  01 2018 Mcalibrator2 was updated so that it can work on the latest MATLAB 2017a


subfunctions/colorimeter/cs100a.m
classdef cs100a
  % a class to manipulate MINOLTA CS-100A from MATLAB through a serial port connection
  %
  % Created    : "2012-04-11 09:23:57 ban"
  % Last Update: "2016-09-13 18:19:02 ban"

  properties (Hidden) %(SetAccess = protected)
    portname='COM1'; % id of serial port to communicate with CS-100A
    rscom=[];  % serial port object
  end

  properties
    init_flg=0;
  end

  methods

    % constructor
    function obj=cs100a(port_name)
      if nargin==1 && ~isempty(port_name)
        obj.portname=port_name;
      end
    end

    % destructor
    function obj=delete(obj)
      if ~isempty(obj.rscom)
        fclose(obj.rscom);
        delete(obj.rscom);
        obj.rscom=[];
        obj.init_flg=0;
      end
    end

    % create/open a serial port connection to communicate with CS-100A
    function obj=gen_port(obj,port_name)
      if nargin>1 && ~isempty(port_name)
        obj.portname=port_name;
        obj.rscom=serial(obj.portname);
      elseif ~isempty(obj.portname)
        obj.rscom=serial(obj.portname);
      else
        error('set a name of serial port.');
      end

      set(obj.rscom,'DataBits',7,'BaudRate',4800,'Parity','even','StopBits',2,'Terminator','CR/LF');

      fopen(obj.rscom);

      % NEW for MATLAB R14
      % After initializing the serial port using FOPEN,
      % you should set the RTS and DTR pins to low using the following code:
      set (obj.rscom, 'DataTerminalReady' ,'off') ;
      set (obj.rscom, 'RequestToSend' ,'off') ;
      % and pause for some msec
      pause(0.3);

      % When the RTS and DTR pins are subsequently set high at any time using the following code,
      % the voltage at the pins will be as expected:
      set (obj.rscom, 'DataTerminalReady' ,'on') ;
      set (obj.rscom, 'RequestToSend' ,'on') ;
      
      obj.init_flg=1;
    end

    % reset a serial port connection
    function obj=reset_port(obj)
      fclose(obj.rscom);
      delete(obj.rscom);
      obj.rscom=[];
      obj.init_flg=0;
    end

    % initialize CS-100A
    function [obj,check,integtime]=initialize(obj,integtime)

      if isempty(obj.rscom), error('serial connection has not been established. run gen_port first.'); end

      % NOTICE 1
      %
      % First, you must confirm electric (electrical) power supply 'OFF'.
      % Then, press 'ON' button of the CS-100A, pushing 'F' key.
      % Check whether the character 'c' was shown on the bottom right corner of the CS-100A display.
      % Through these process, CS-100A will be set as bothway communications mode.

      % NOTICE 2
      % 'integtime' is a dummy variable to match the format with the other class.
      % It is not used in this function.

      fprintf(obj.rscom, 'MDS,07'); % Set Up Measurement Parameters.
      pause(0.1);

      % Checking setting parameters.
      results=fgets(obj.rscom);

      if ~strcmp(results(1:4),'OK00')
        disp('setup error. check CS-100A and cable connections.');
        return;
      end

      % Format -MDS
      %             'MDS,[][]'
      %               00 : MINOLTA Standard Calibration Mode
      %               01 : Optional Calibration Mode
      %               04 : Chromaticity Measurement Mode
      %               05 : Color Difference Measurement Mode
      %               06 : Measurement Response Time --- 100ms Fast Mode
      %               07 : Measurement Response Time --- 400ms Slow Mode

      % Error Check Command Description
      %               ER00 : Command Error, or Parameters Error
      %               ER11 : Memory Values Error
      %               ER20 : EEPROM Error
      %               ER30 : Battely Out Error

      fprintf(obj.rscom, 'MDS,04'); % Set Up Measurement Parameters.
      pause(0.1);

      % Checking setting parameters.
      results=fgets(obj.rscom);

      if strcmp(results(1:4),'OK00')
        check = 0;
      else
        check = 1;
      end
    end

    % measure CIE1931 xyY of the target
    function [qq,Y,x,y,obj]=measure(obj,integtime)

      if isempty(obj.rscom), error('serial connection has not been established. run gen_serial first.'); end
      if nargin<2 || isempty(integtime), integtime=500; end %#ok

      % NOTICE 1
      % Terminator (delimiter) should be CR/LF (CR+LF).
      % This terminator was set by CreateSerialObjectMINOL.

      % NOTICE 2
      % 'integtime' is a dummy variable to match the format with the other class.
      % It is not used in this function.

      fprintf(obj.rscom, 'MES'); % Measure Light Under Conditions defined by the 'MDS' command line.
      pause(0.1);

      % Format -MES
      %       output : normal
      %                   'OK00,+[][][][][][],+.[][][][],+.[][][][]'
      %                     notice : here, + may chage to -
      %              : luminance and chromaticity over the measurement range
      %                   'ER10'
      %              : chromaticity over the measurement range
      %                   'OK11,+[][][][][][],+.0000,+.0000'
      %                     notice : here, + may chage to -
      %              : luminance over the measurement range
      %                   'OK12,-999999,+.[][][][],+.[][][][]'
      %                     notice : here, + may chage to -
      %              : values are over the display or maasurement range
      %                   'ER12'
      %              : luminance under the masurement range
      %                   'OK13,+[][][][][][],+.[][][][],+.[][][][]'
      %                     notice : here, + may chage to -

      results=fscanf(obj.rscom);
      %if strcmp(results(1:4),'OK00')
        qq=0;
        Y=str2num(results(6:12)); %#ok
        x=str2num(results(14:19)); %#ok
        y=str2num(results(21:26)); %#ok
        return
      %end

      % numretry=1; qq=1;
      % while numretry<=5 && qq~=0
      %   numretry=numretry+1;
      %   if strcmp(results(1:4),'OK00')
      %     Y=str2num(results(6:12)); %#ok
      %     x=str2num(results(14:19)); %#ok
      %     y=str2num(results(21:26)); %#ok
      %     qq=0;
      %   else
      %     if strcmp(results(1:4),'OK12')
      %       fprintf(obj.rscom, 'MDS,06'); % fast mode
      %       fscanf(obj.rscom); % clear data
      %     elseif strcmp(results(1:4),'OK13')
      %       fprintf(obj.rscom, 'MDS,07'); % slow mode
      %       fscanf(obj.rscom); % clear data
      %     end
      %     pause(0.1);
      %     fprintf(obj.rscom, 'MES');
      %     pause(0.1);
      %     results=fscanf(obj.rscom);
      %   end
      % end
      % 
      % if numretry>5
      %   Y=str2num(results(6:12)); %#ok
      %   x=str2num(results(14:19)); %#ok
      %   y=str2num(results(21:26)); %#ok
      % end

      % Descriptions
      % qq                  : Measurement quality code (OK00 = O.K.)
      % U                   : Unit of Measured Luminace, always 0 with CS-100A
      % Y                   : +(-)[][][][][][] 1931 CIE Y (units indicated by U)
      % x                   : +(-).[][][][] 1931 CIE x
      % y                   : +(-).[][][][] 1931 CIE y
    end

  end % methods

end % classdef cs100a
